### DISCUSSION TITLE
Time Zones

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/webwork2#1528] [General]
As I manage the Runestone server, where multiple schools across the globe have courses running, I'm hitting some issues with time zones that you wouldn't normally encounter in a more typical installation. The server's time zone is UTC. And WeBWorK's default time zone is also UTC. But each course sets its own time zone/locale in the course level config file. I have the issues below and am posting to solicit input, objections, etc.

1. The answer.log file is recording UTC times. I'm unsure (yet) if it is using the server time zone or the site.conf time zone. To an instructor, oblivious to anything but their own time zone, this is confusing. I'd like to change things so that each line of the answer.log explicitly states the time zone/locale it is using. Would this cause problems? Is there anything that is relying on the current structure of each line of the answer.log file?
2. I'm considering changing the answer.log file to use the course's actual time zone (instead of the server time zone or the site.conf time zone). Should that be up for discussion?
3. The transaction logs in `\opt\webwork\webwork2` are using UTC (again, not sure at the moment if it's the server time zone or the site.conf time zone). I think this should be the server time zone. So if I look into it and it's actually using the site.conf time zone, may I change it?
4. The "Start Date" on a Gateway Quiz version that a user sees on the Homework Sets page is using UTC. I think this should definitely change to use the course's time zone. Are there any objections?

**Comment by drgrice1** [comment]
I agree that the answer.log should use the course time zone.  The other course log file should do the same.

I don't agree that the general WeBWorK transaction log file should use the server time zone instead of the site.conf setting.

I agree on 4.

**Comment by Alex-Jordan** [comment]
My thinking for the dates in the transaction log is that this is something only a server admin will access. So for starters, it's for someone who knows more what they are looking at. But also, the timezone/locale in site.conf can be changed just by an edit to site.conf. So with the flick of a switch, prior log data becomes incorrect by some shift. Such a shift is less likely to happen if it's the server's time that is used. Can you elaborate more on why to prefer the site.conf time?

I had forgotten about each course's login log. That too is currently using UTC on my server. (I have not yet looked into whether it is using the server's time, or shifting to the site.conf timeone locale which happens to be the same.)

**Comment by drgrice1** [comment]
The site.conf is also something that only a server admin will access.  I think if the server admin sets the time zone in the site.conf file, then that is what should be used.  It is just as easy to change the servers time zone for an admin as it is to change there site.conf setting.  Usually these will probably be the same (at least that is how my servers are set up, and how I would do it). I feel that the site.conf seeing is what WeBWorK should use to keep WeBWorK configuration easy and in the purview of the config files as much as possible.

**Comment by Alex-Jordan** [comment]
Fair enough. Now to discern what is actually the current state of affairs.

I opened a PR for item 4. I'd like to open one soon for the answer.log and login.log to switch to use the course's time.

If I can be thorough, I'll see if the transaction log is using site.conf time or server time, and if there's a need to change it.

**Comment by taniwallach** [comment]
I also agree that it would be desirable that course log files (`login.log` and `answer.log`) use the course time-zone.

About listing the time-zone in the log messages, it would remove ambiguity, which is probably worth the extra data being included on each   line of the file for any site which hosts courses from multiple time zones. As such, I am supportive of having an option to include the time-zone in the log messages. 

We could consider using a variable set in `site.conf` (which can be changed at the course level) to control whether or not to include the time-zone in the log message "timestamp". (I think it can default to off.) `writeCourseLog` from `lib/WeBWorK/Utils.pm` is passed a  course-environment, so getting the course-level value of such a setting should be feasible. The downside is that there would be some performance overhead to checking the switch for every log call.

I am not aware of any "WeBWorK Project" code which makes use of the records stored in `answer.log` or `login.log`. It is quite likely that some local scripts to process these file exist at some institutions. If so a format change could require fixing the scripts.

About how to provide the time-zone in log files,  I recommend providing the "time-zone" in a format which can be parsed back into a correct time. Thus, in my opinion using `%Z` should be avoided. I would suggest considering one of:
  - long time-zone string via "%{time_zone_long_name}", or 
  - a UTC-offset via "%z".

Background: Short time zone abbreviations (like `IST` and `IDT`) are ambiguous in a worldwide context, and many do not work when Perl attempts to parse them back into "real" time-zone name (as they are not unique). That issue led to the changes made in https://github.com/openwebwork/webwork2/pull/900 and https://github.com/openwebwork/webwork2/pull/941 . See also the discussions in https://github.com/openwebwork/webwork2/issues/928 .

**Reply by dlglin** [reply]
I don't think the variable should be set in `site.conf`.  Only settings that can not have a default value because they are specific to the particular installation belong in `site.conf`.  This should go in `defaults.config` with the option to override elsewhere.
